hazelcast-client-new/src/main/java/com/hazelcast/client/txn/proxy/xa/XAResourceProxy.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.txn.proxy.xa;
+
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.client.impl.protocol.parameters.DataCollectionResultParameters;
+import com.hazelcast.client.impl.protocol.parameters.XACollectTransactionsParameters;
+import com.hazelcast.client.impl.protocol.parameters.XAFinalizeTransactionParameters;
+import com.hazelcast.client.spi.ClientContext;
+import com.hazelcast.client.spi.ClientProxy;
+import com.hazelcast.client.spi.ClientTransactionManagerService;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.nio.serialization.SerializationService;
+import com.hazelcast.transaction.HazelcastXAResource;
+import com.hazelcast.transaction.TransactionContext;
+import com.hazelcast.transaction.TransactionOptions;
+import com.hazelcast.transaction.impl.xa.SerializableXID;
+import com.hazelcast.transaction.impl.xa.XAResourceImpl;
+
+import javax.transaction.xa.XAException;
+import javax.transaction.xa.XAResource;
+import javax.transaction.xa.Xid;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Client side XaResource implementation
+ */
+public class XAResourceProxy extends ClientProxy implements HazelcastXAResource {
+
+    private static final int DEFAULT_TIMEOUT = (int) TimeUnit.MILLISECONDS.toSeconds(TransactionOptions.DEFAULT_TIMEOUT_MILLIS);
+    private final ConcurrentMap<Long, TransactionContext> threadContextMap = new ConcurrentHashMap<Long, TransactionContext>();
+    private final ConcurrentMap<Xid, List<TransactionContext>> xidContextMap
+            = new ConcurrentHashMap<Xid, List<TransactionContext>>();
+
+    public XAResourceProxy(String serviceName, String objectName) {
+        super(serviceName, objectName);
+    }
+
+    @Override
+    public TransactionContext getTransactionContext() {
+        long threadId = Thread.currentThread().getId();
+        return threadContextMap.get(threadId);
+    }
+
+    @Override
+    public synchronized void start(Xid xid, int flags) throws XAException {
```

Comment:
I didn't quite understand the internals here, but is it really required to make all these methods `synchronized`?